### PR TITLE
Enable cross build

### PR DESCRIPTION
**Pull request by mkumatag**
This PR enables the cross building both binary and docker images.

### CODE REVIEW

**Comment by deitch** [comment]
Overall, looks pretty good. Captures everything. right down to `ci` and `cd` targets. :-)

Some comments:

* why the use of `CALICO_UPGRADE_CONTAINER_NAME` var in Makefile and not just `CONTAINER_NAME` like all of the other projects? The more we unify, the more we can look at shared makefile infrastructure
* why is `calico-upgrade-$(ARCH)` in `bin/` but `kubectl` is in `dist/$(ARCH)/kubectl`?
* [here](https://github.com/projectcalico/calico-upgrade/pull/40/files#diff-b67911656ef5d18c4ae36cb6741b7965R197) could we keep it the same target structure as the `upgrade` binary target, and have (combining both), unless you see a reason otherwise:

```makefile
 bin/kubectl: bin/kubectl-$(ARCH)

bin/kubectl-$(ARCH):
	$(CURL) -L $(KUBECTL_URL) -o - | tar -zxvf - --strip-components=3 -O kubernetes/client/bin/kubectl > $@
	chmod +x $@
```

**Comment by mkumatag** [comment]
> why the use of CALICO_UPGRADE_CONTAINER_NAME var in Makefile and not just CONTAINER_NAME like all of the other projects? The more we unify, the more we can look at shared makefile infrastructure
> why is calico-upgrade-$(ARCH) in bin/ but kubectl is in dist/$(ARCH)/kubectl?
> here could we keep it the same target structure as the upgrade binary target, and have (combining both), unless you see a reason otherwise:

I really don't have a good reason why these files are in different location, lemme combine them together.

**Comment by mkumatag** [comment]
@deitch fixed the review comments and with that removed those `chmod` statements from Dockerfiles which are not really necessary and allow to build these dockerimages without qemu.

**Comment by deitch** [comment]
Thanks @mkumatag . The cross-build part looks better and more standardized.

The other part I am trying to wrap my head around is the `ci` targets. As far as I can tell, Semaphore currently executes the following (ignoring just basic version checking and such):

1. `make semaphore`
2. complex script which could be replaced by `make cd`

Our goal throughout has been to replace that with `make ci` where possible. It looks like `make ci` and `make cd` are not quite unified, so that we can change `make semaphore` to `make ci`. They both execute `clean` and `st` targets, `ci` also executes `static-checks` (which is good), and `semaphore` also clears out some junk [here](https://github.com/mkumatag/calico-upgrade/blob/246235e708cd477c0fb0d227bb1a8a5cbe1e6f04/Makefile#L356) (@fasaxc do you know what this does?) and builds the image and binaries.

Here is my suggestion, but go ahead and suggest something else if this doesn't make sense.

1. add `image` to `ci`
2. add the `build` target you put [here](https://github.com/mkumatag/calico-upgrade/blob/246235e708cd477c0fb0d227bb1a8a5cbe1e6f04/Makefile#L366) to `ci`
3. Change `semaphore` target to just depend on `ci`

If this all works, then @fasaxc can change Semaphore for this to just do `make ci` and `make cd` like he did elsewhere, and then another PR can remove the `semaphore` target entirely, thus unifying this project with the others.

I think I might have missed something with that cleanup section [above](https://github.com/mkumatag/calico-upgrade/blob/246235e708cd477c0fb0d227bb1a8a5cbe1e6f04/Makefile#L356), but not sure what it does.

Thoughts?

**Comment by mkumatag** [comment]
> Here is my suggestion, but go ahead and suggest something else if this doesn't make sense.
> 
> add image to ci
> add the build target you put here to ci
> Change semaphore target to just depend on ci

Make sense to me, addressed all of them.

**Comment by deitch** [comment]
That appears to cover them all. I think we should let @fasaxc / @caseydavenport / @tomdee comment here now. I don't know what that additional line does or if it is necessary.

**Comment by mkumatag** [comment]
ping @fasaxc @caseydavenport @tomdee

**Comment by fasaxc** [comment]
I think the build included a windows binary before, have we lost that?

**Comment by mkumatag** [comment]
> I think the build included a windows binary before, have we lost that?

Yes, you are right, I see them building in `dist/*` target but they are never shipped or built any image with them. Do we really need those binaries?

**Comment by fasaxc** [comment]
The binaries are shipped via github; for example: https://github.com/projectcalico/calico-upgrade/releases/tag/v1.0.5

**Comment by mkumatag** [comment]
> The binaries are shipped via github; for example: https://github.com/projectcalico/calico-upgrade/releases/tag/v1.0.5

Give me some time, will accommodate the changes.

**Comment by mkumatag** [comment]
@fasaxc I refactored overall code, now you will see following binaries in the `dist` folder and will help in pushing other arch images as well:

```shell
root@localhost:~/calico-upgrade# ls -ltrh dist/calico-upgrade*
-rwxr-xr-x 1 root root 30M Jul 17 10:06 dist/calico-upgrade-linux-s390x
-rwxr-xr-x 1 root root 28M Jul 17 10:06 dist/calico-upgrade-linux-arm64
-rwxr-xr-x 1 root root 29M Jul 17 10:06 dist/calico-upgrade-linux-ppc64le
-rwxr-xr-x 1 root root 34M Jul 17 10:06 dist/calico-upgrade-darwin-amd64
-rwxr-xr-x 1 root root 30M Jul 17 10:10 dist/calico-upgrade-windows-amd64.exe
-rwxr-xr-x 1 root root 29M Jul 17 10:32 dist/calico-upgrade


root@localhost:~/calico-upgrade# file dist/calico-upgrade*
dist/calico-upgrade:                   ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, stripped
dist/calico-upgrade-darwin-amd64:      Mach-O 64-bit x86_64 executable
dist/calico-upgrade-linux-arm64:       ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, stripped
dist/calico-upgrade-linux-ppc64le:     ELF 64-bit LSB executable, 64-bit PowerPC or cisco 7500, version 1 (SYSV), statically linked, stripped
dist/calico-upgrade-linux-s390x:       ELF 64-bit MSB executable, IBM S/390, version 1 (SYSV), statically linked, stripped
dist/calico-upgrade-windows-amd64.exe: PE32+ executable (console) x86-64 (stripped to external PDB), for MS Windows
root@localhost:~/calico-upgrade#
```

File: pkg/commands/report.go [hunk 1]
```diff
@@ -62,7 +62,7 @@ func outputCheckPermsAndExit(output string) {
 	ch.Separator()
 	ch.Error("Unable to prepare the reports directory for writing the migration reports. " +
 		"Please ensure your file permissions allow the following:")
-	ch.Bullet(fmt.Sprintf("creation of the output directory if it doesn't already exist " +
+	ch.Bullet(fmt.Sprintf("creation of the output directory if it doesn't already exist "+
```
Comment (fasaxc) [inline pkg/commands/report.go]: Random change?

File: Makefile [hunk 2]
```diff
@@ -1,13 +1,59 @@
...
+ifeq ($(ARCH),x86_64)
+    override ARCH=amd64
+endif
 
 ###############################################################################
 # Go Build versions
 GO_BUILD_VER:=v0.16
 CALICO_BUILD?=calico/go-build:$(GO_BUILD_VER)
 
+# Figure out the users UID.  This is needed to run docker containers
```
Comment (fasaxc) [inline Makefile]: Typo: `s/users/user's/`

File: pkg/commands/report.go (hunk 1 above)
Comment (mkumatag) [reply pkg/commands/report.go]: while checking `static-checks` there was a failure related to fmt, so fixed it.

File: Makefile (hunk 2 above)
Comment (mkumatag) [reply Makefile]: sure, will fix it.

**Comment by mkumatag** [comment]
@fasaxc things are working fine - https://gist.github.com/mkumatag/00ac665be7f6481a32f9d9d791091024

**Comment by tomdee** [comment]
I'm also seeing problems when I run `make ci`

**Comment by mkumatag** [comment]
This seems like a know issue with `-i` flag with go build per https://github.com/golang/go/issues/26071, I just removed `-i` option from `go build`, please rerun the build and check if you are still facing this issue.

**Comment by mkumatag** [comment]
@tomdee @fasaxc did you get a chance to rerun this PR?

**Comment by caseydavenport** [comment]
This seems to have gotten lost... I'm going to merge it :D

**Pull request merged**